opte/src/engine/ip6.rs

```diff
@@ -435,7 +447,7 @@ impl<'a> Ipv6Hdr<'a> {
     /// Return the total length of the packet, including the base header, any
     /// extension headers, and the payload itself.
     pub fn total_len(&self) -> usize {
-        usize::from(self.base.payload_len()) + Self::BASE_SIZE
+        self.pay_len() + Self::BASE_SIZE
```

Comment:
Should this also be updated to `self.hdr_len() + self.pay_len()`?

Reply:
I don't think so. `Ipv6Hdr::hdr_len()` returns the length _including the extension headers_. To be clear, I am not sure that it _should_ return that, but that's what it currently does and what the documentation claims that it does.

Reply:
Ah ok, so that's a bit different than the methods on `Ipv6Meta` which is a bit confusing. Would it be too much to update one of the other so we're consistent

Reply:
I agree that it's all extremely confusing. That said, I see `Ipv6Meta::hdr_len` as returning `BASE_SIZE + self.ext_len`, which I think is consistent with `Ipv6Hdr`, no?

Reply:
I see part of the issue. The `Ipv6Meta::total_len()` method appears incorrect as well. That returns `self.hdr_len() + self.pay_len`. But `Ipv6Meta::hdr_len()` should return the `BASE_SIZE + self.ext_len`. We are currently using the `pay_len` field to set the actual Payload Length field of the IPv6 base header, which implies that it includes the length of the extension headers. Those two things can't both be true: either `pay_len` includes the extension headers, and the `Ipv6Meta::total_len()` method is wrong; or `pay_len` does _not_ include the extension headers, and we're setting the IPv6 Payload Length field incorrectly.

Reply:
I should say, part of the confusion comes from the fact that we have _similar_ methods for IPv4 and IPv6. I say similar, because they're just named the same thing, they are not trait methods or anything statically constrained to be the same by the compiler.

That's a problem, because the protocols are so different that they don't really have the same concepts. It's not clear what you mean when you say "header length" or "payload length" in the case of IPv6. We have methods that return all of the different pieces (base header length, base + extensions, just extensions, IPv6 Payload Length == extensions + ULP, and ULP length), but it's not clear to me they're being used correctly.

Reply:
Alright, I've found the issue. I'm going to push that onto this PR, since it's pretty small and really all related. @luqmana Let me know if you think there is more cleanup or documentation to be added around the various IPv6 methods. It's a bit confusing, but I believe it's now consistent and I've added some more documentation.